x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlDailyMaintenanceService.java

```diff
@@ -136,32 +149,127 @@ private void triggerTasks() {
-            auditUnassignedMlTasks(clusterService.state());
+
+            // Step 3: Log whether the tasks have finished
+            ActionListener<AcknowledgedResponse> finalListener = ActionListener.wrap(
+                unused -> LOGGER.info("Completed [ML] maintenance tasks"),
+                e -> LOGGER.error("An error occurred during maintenance tasks execution", e)
+            );
+
+            // Step 2: Delete expired data
+            ActionListener<AcknowledgedResponse> deleteJobsListener = ActionListener.wrap(
+                unused -> triggerDeleteExpiredDataTask(finalListener),
+                e -> {
+                    LOGGER.info("[ML] maintenance task: triggerDeleteJobsInStateDeletingWithoutDeletionTask failed", e);
+                    // Note: Steps 1 and 2 are independent of each other and step 2 is executed even if step 1 failed.
+                    triggerDeleteExpiredDataTask(finalListener);
+                }
+            );
+
+            // Step 1: Delete jobs that are in deleting state
+            triggerDeleteJobsInStateDeletingWithoutDeletionTask(deleteJobsListener);
+
+            auditUnassignedMlTasks();
         } finally {
             scheduleNext();
         }
     }
 
+    private void triggerDeleteExpiredDataTask(ActionListener<AcknowledgedResponse> finalListener) {
+        ActionListener<DeleteExpiredDataAction.Response> deleteExpiredDataActionListener = ActionListener.wrap(
+            deleteExpiredDataResponse -> {
+                if (deleteExpiredDataResponse.isDeleted()) {
+                    LOGGER.info("Successfully completed [ML] maintenance task: triggerDeleteExpiredDataTask");
+                } else {
+                    LOGGER.info("Halting [ML] maintenance tasks before completion as elapsed time is too great");
+                }
+                finalListener.onResponse(new AcknowledgedResponse(true));
+            },
+            finalListener::onFailure
+        );
+
+        executeAsyncWithOrigin(
+            client,
+            ML_ORIGIN,
+            DeleteExpiredDataAction.INSTANCE,
+            new DeleteExpiredDataAction.Request(deleteExpiredDataRequestsPerSecond, TimeValue.timeValueHours(8)),
+            deleteExpiredDataActionListener);
+    }
+
+    // Visible for testing
+    public void triggerDeleteJobsInStateDeletingWithoutDeletionTask(ActionListener<AcknowledgedResponse> finalListener) {
+        SetOnce<Set<String>> jobsInStateDeletingHolder = new SetOnce<>();
+
+        ActionListener<List<AcknowledgedResponse>> deleteJobsActionListener = ActionListener.wrap(
+            deleteJobsResponses -> {
+                if (deleteJobsResponses.stream().allMatch(AcknowledgedResponse::isAcknowledged)) {
+                    LOGGER.info("Successfully completed [ML] maintenance task: triggerDeleteJobsInStateDeletingWithoutDeletionTask");
+                } else {
+                    LOGGER.info("At least one of the ML jobs could not be deleted");
```

Comment:
If we ever see this in a log it will frustrate us.  We will want to know which ones failed.  The low level error logging for the failure might be on a different node in the cluster that we haven't been sent the logs for.  So this should list the job IDs that failed to be deleted.